Adding tests to 'contacts' app

DanielMy777 opened the pull request:
- Adding tests via pytest[1] to the contacts app, focusing on tests for the
connection model.
- Changing the Apartment model import in the test data file.
List of tests is stated in the commit log.
fix #64 
[1]: https://docs.pytest.org/en/6.2.x/

Micha546 commented on contacts/tests.py:
```diff
@@ -1,3 +1,122 @@
+        num_of_rooms=rooms,
+        start_date=s_date
+    )
+    return apartmant
```
Why did you add all those create functions? what is their purpose?

DanielMy777 replied:
There's a strong purpose,
I have to create multiple objects in order to create connections.
Those include creating an apartment, seeker, and each requires creating a user too.

1Oreo commented:
@DanielMy777 since this is a testing PR I would suggest you rebase it so the purest can run and we can see in the PR that the tests work as intended.

Micha546 commented on contacts/tests.py:
```diff
@@ -1,3 +1,122 @@
+
+
+@pytest.mark.django_db
+def create_user(email, f_name, l_name, b_date, passw):
```
create_user is already a function in User, maybe change the name to avoid a potential conflict

DanielMy777 replied:
Noted, I'll change the name

DanielMy777 commented:
@1Oreo I've rebased, fixed some stuff that were hurt during the last merges, and the tests run great now.
Waiting for your further review :)

tamirho commented:
LGTM

OrAsaf commented on contacts/tests.py:
```diff
@@ -1,3 +1,123 @@
+
+
+@pytest.mark.django_db
+def test_invalid_duplicate_connection(connection_1, seeker_1, apartment_1):
+    with pytest.raises(IntegrityError):
+        with transaction.atomic():
+            Connection(seeker=seeker_1, apartment=apartment_1).save()
```
This is a bit ambiguous. You send connection_1 to the function but don't use it, I think I know why you are doing just that but I would like to hear the reason please.

DanielMy777 replied:
@1Oreo Connection_1 is created as a pytest fixture, it saves a row in the db with a connection between seeker_1 and apartment_1.
In the test, I want to save *another* connection row between seeker_1 and apartment_1, which should cause an IntegrityError. To make that first original connection be inserted to the db I had to send connection_1 to the function. It's a fixture so as long as I don't add it as an argument, it's like that row was never added in the first place (will not activate the fixture and therefore will not save connection_1). Causing the "second" row that was added in the test function to be the first and only one, which will not cause a duplication error.

OrAsaf replied:
So I figured this is what you meant to create.
But the way you implemented it might cause someone to misunderstand the intention. What I would do to sort it out is use `connection_1.seeker` and `connection_1.appartment` so that when someone reads the code he will understand that you are re-inserting the same connection to cause an error.

DanielMy777 replied:
@1Oreo You got me 🎯 
I took your tip and changed it, let me know what you think :)

Omeramsc commented on contacts/tests.py:
```diff
@@ -1,3 +1,123 @@
+
+
+@pytest.fixture
+def seeker_2(new_city):
```
You can achieve the same using one fixture Instead of creating two, almost identical ones
for example: 
```
@pytest.fixture
def fixture_with_different_results():
    def _inner_function(par_1, par_2, par_3):
        changing_list = [par_1,par_2,par_3]
        return changing_list

    return _inner_function

def test_the_fixture(fixture_with_different_results):
    first_item = fixture_with_different_results(1,2,3)
    second_item = fixture_with_different_results(4,5,6)
```

first_item will get a list containing [1,2,3]
second_item will get a list containing [4,5,6]

Omeramsc replied:
It'll also give you more flexibility with the information you want to insert for the test.

DanielMy777 replied:
@Omeramsc Wouldn't that require me to call this fixture and create a seeker inside the test function?
I thought that the main point of a fixture is to have blocks of code happen before the test starts, that's what I got from Edward. For example, if I understand correctly you'd like something like:
```
def test_invalid_users_connection(fixture_with_different_results):
    seeker_1 = fixture_with_different_results(*seeker1 data*)
    seeker_2 = fixture_with_different_results(*seeker2 data*)
    with pytest.raises(ValueError):
        with transaction.atomic():
            Connection(seeker=seeker_1, apartment=seeker_2).save()
```
There's other rows I neglected as well (creating the user, which is part of the seeker data, before calling the fixture), Which will add lines of code and complexity IMO.

Omeramsc replied:
Shorter isn't always better or less complex; While this feature of fixtures is great, I personally wouldn't create two separate fixture for entries as similar as those. @simzacks @1Oreo  WDYT?

OrAsaf replied:
I agree with what @Omeramsc said. With the way the project is now (small and in the early days) writing a fixture for every fixture can work, and still be readable and understandable, but in a much bigger application it will again create ambiguity. One of the traits good code needs is to be readable but also understandable and I think what Omer suggested shows someone who reads the code that the fixture is of the same type, just with 2 different users, and also "saves" duplicating code.

DanielMy777 replied:
@Omeramsc I've never seen this and couldn't really find it online, but gave it a shot.
I changed to code, please tell me if this was what you meant, Thanks!

Omeramsc replied:
This is called factories as fixtures, you can read more about it (and pytest fixtures) here:
https://docs.pytest.org/en/stable/fixture.html#factories-as-fixtures

I left some comments, but it already looks way better. good job. :)

DanielMy777 replied:
@Omeramsc Thanks! Glad to know that this looks better.
I fixed the code due to your comments, please review again and let me know if there's anything else I should change.

DanielMy777 replied:
Also, pytest won't work until #91  & #92  are merged so please ignore it while reviewing.

Omeramsc commented on contacts/tests.py:
```diff
@@ -1,3 +1,123 @@
+
+
+@pytest.fixture
+def apartment_1(new_city):
```
If there's no `apartment_2` or something of that sort (and applying my former comment should remove the need for those), please name the fixtures in a better way (this comment applies to all of them).

DanielMy777 replied:
Thanks, I renamed the fixtures and marking this as resolved!

Omeramsc commented on contacts/tests.py:
```diff
@@ -1,3 +1,123 @@
+
+
+@pytest.mark.django_db
+def create_new_user(email, f_name, l_name, b_date, passw):
```
You can apply the fixture logic I mentioned for those functions as well.

DanielMy777 replied:
As these aren't fixtures, I'm having trouble understanding what you mean. I've tried some things but to no avail, can you please give me an example?
(The fixture logic applied above uses these functions to create objects in the models)

Omeramsc replied:
I meant that if you're already going the fixture road, you can make `create_new_user` into a fixture as well. fixtures can use other fixtures, so you should not have issues applying it in the current situation.

DanielMy777 committed:
Adding tests to 'contacts' app

Tests include:
- Creating a new connection.
- Accessing an existiong connection.
- Creating multipile connections.
- Failing to create a duplicate connection.
- Failing to create a connection that is not between
  a seeker and an apartment.

Omeramsc commented on contacts/tests.py:
```diff
@@ -1,3 +1,125 @@
+
+
+@pytest.fixture
+def fixture_make_seeker(new_city):
```
Please refrain from mentioning "fixture" in the name, as it is already declared in the decorator. 
a function's name is important and has a huge role in readability, so always pay attention to it.

DanielMy777 replied:
I removed the 'fixture' part and called the function 'make_seeker'

Omeramsc commented on contacts/tests.py:
```diff
@@ -1,3 +1,125 @@
+
+@pytest.fixture
+def fixture_make_seeker(new_city):
+    def make_seeker(user, start_date, min_rent, max_rent, roomates, rooms, about):
```
_make_seeker*

DanielMy777 replied:
I renamed the inner function '_make_seeker' as you suggested

Omeramsc commented on contacts/tests.py:
```diff
@@ -1,3 +1,125 @@
+        seeker.save()
+        return seeker
+
+    return make_seeker
```
_make_seeker*

Omeramsc commented on contacts/tests.py:
```diff
@@ -1,3 +1,126 @@
+    return city
+
+
+def create_seeker(user, city, s_date, min_r, max_r, roomates, rooms, about):
```
This can be included in the fixture - no need for a separate function.

DanielMy777 replied:
Included in a fixture

Omeramsc commented on contacts/tests.py:
```diff
@@ -1,3 +1,126 @@
+    return user1
+
+
+def create_city():
```
We've got a fixture doing the exact same thing, and this is never getting used, you can remove it.

DanielMy777 replied:
You're right it was old, didn't notice. Removed.

Omeramsc commented on contacts/tests.py:
```diff
@@ -1,3 +1,126 @@
+    return seeker
+
+
+def create_apartment(user, city, addr, rent, roomates, rooms, s_date):
```
same as above

DanielMy777 replied:
Included in a fixture.

DanielMy777 commented:
@Omeramsc I think I finally got what you meant now, please review my changes and let me know what you think :)

Omeramsc reviewed (approved):
Looking much better! I'm glad to see you got it.
